accumulator/batch.go

```diff
@@ -86,28 +87,43 @@ func (b *Batch) RegisterInvocation(
 ) {
 	b.mu.Lock()
 	defer b.mu.Unlock()
-	b.invocations[requestID] = &Invocation{
+	inc := &Invocation{
 		RequestID:   requestID,
 		FunctionARN: functionARN,
 		DeadlineMs:  deadlineMs,
 		Timestamp:   timestamp,
 	}
+	if w, ok := b.invocations[waitingInvocationKey]; ok {
+		// If the agent payload is cached before the invocation is registered
+		// then associate the agent payload with the invocation.
+		inc.AgentPayload = w.AgentPayload
+		inc.TransactionID = w.TransactionID
+		delete(b.invocations, waitingInvocationKey)
+	}
+	b.invocations[requestID] = inc
 	b.currentlyExecutingRequestID = requestID
 }
 
 // OnAgentInit caches the transactionID and the payload for the currently
 // executing invocation as reported by the agent. The traceID and transactionID
 // will be used to create a new transaction in an event the actual transaction
-// is not reported by the agent due to unexpected termination.
+// is not reported by the agent due to unexpected termination. If the current
+// invocation is not registered yet then the payload is cached temporarily and
+// associated with the invocation when it is registered.
 func (b *Batch) OnAgentInit(transactionID string, payload []byte) error {
+	if !isTransactionEvent(payload) {
+		return errors.New("invalid payload")
+	}
 	b.mu.Lock()
 	defer b.mu.Unlock()
 	i, ok := b.invocations[b.currentlyExecutingRequestID]
 	if !ok {
-		return fmt.Errorf("invocation for requestID %s does not exist", b.currentlyExecutingRequestID)
-	}
-	if !isTransactionEvent(payload) {
-		return errors.New("invalid payload")
+		// It is possible that the invocation is registered at a later time
+		b.invocations[waitingInvocationKey] = &Invocation{
```

Comment:
Should we require the agent to send the Lambda request ID along with the transaction ID and payload? Then it can always match up on request ID, and remove currentlyExecutingRequestID.

Reply:
That sounds plausible. I was motivated by keeping the lifecycle simple by exposing less arguments to the agent but I think sending requestID (maybe as a header?) will be better. Let me make the adjustments.

Reply:
I have updated the PR to accept request ID as a header `x-elastic-aws-request-id`. I didn't remove `currentlyExecutingRequestID` because I think adding the constraint for all agent data to have request ID as a header might not be a good idea but adding the requirement to the txn registration API should be okay. Additionally, the `currentlyExecutingRequestID` can be set reliably in both agent init or register invocation stages to ensure a correct value for later calls. Let me know your thoughts on it.

Reply:
Right, I forgot that we still need to associate events with the invocation. Looks good, thanks.